### ISSUE TITLE
[typescript-angular2] Arrays in query params not using preferred format

### ISSUE THREAD

**Issue by RichiWIP** [swagger-api/swagger-codegen#4531]
##### Description

I am currently developing a middleware API and I generate my angular2 client using codegen. I have one call that expects a short array of strings as a query parameter. The current generator will create a call like this:
`http://www.foo.bar/somthing?array=value1,value2,value3`

My middleware doesn't understand this format and as far as I have researched, it is preferred to use the following format:
`http://www.foo.bar/somthing?array=value1&array=value2&array=value3`

##### Swagger-codegen version

2.0.17

##### Suggest a Fix

I suggest replacing the current template for queryParams with this code:

``` mustache
{{#queryParams}}
        if ({{paramName}} !== undefined) {
            {{#isListContainer}}
                if({{paramName}}){
                    {{paramName}}.forEach((element) => {
                        queryParameters.append('{{baseName}}', <any>element);
                    })
                }
            {{/isListContainer}}
            {{^isListContainer}}
                queryParameters.set('{{baseName}}', <any>{{paramName}});
            {{/isListContainer}}
        }
{{/queryParams}}
```

**Comment by wing328** [comment]
The `collectionFormat` attribute in the OpenAPI/Swagger 2.0 spec is related to the issue you reported.

Your solution will work for your scenario (collectionFormat=`multi`) but not other valid collectionFormat.  

Would you have time to contribute a solution that works with all supported collection formats?

**Comment by RichiWIP** [comment]
Ah, yes, I see the Problem. A better solution for now would be
```
{{#queryParams}}
        if ({{paramName}} !== undefined) {
            {{#isCollectionFormatMulti}}
                if({{paramName}}){
                    {{paramName}}.forEach((element) => {
                        queryParameters.append('{{baseName}}', <any>element);
                    })
                }
            {{/isCollectionFormatMulti}}
            {{^isCollectionFormatMulti}}
                queryParameters.set('{{baseName}}', <any>{{paramName}});
            {{/isCollectionFormatMulti}}
        }
{{/queryParams}}
```

That will at least support "multi" and "csv" (as ts / js will serialize the array as csv). I will give it a try to add support for the other formats and also add this to form params where allowed.

**Comment by wing328** [comment]
@RichiWIP thanks. For PHP, we use a function to handle different collection format: https://github.com/swagger-api/swagger-codegen/blob/1ea9865a4498e26bb1f4ecf490aaad19d19357a8/samples/client/petstore/php/SwaggerClient-php/lib/ObjectSerializer.php#L173-L205

and use the `collectionFormat` mustache tag to access the `collectionFormat` value in the template:

https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/resources/php/api.mustache#L180-L184

**Comment by RichiWIP** [comment]
I created a gist with my proposal for the template changes. I have added the proper serialization for the different collection formats to query params, form params and header params (here without multi as it is not supported). 
I have also added a constant array of all APIS to apis.mustache to make it easier to import libraries with a great number of api classes. That is the only change to apis.mustache, it is not required for the actual implementation of collection formats.

https://gist.github.com/RichiWIP/8bd23661a4af9629f4c47eb8ab8d25ad
